Initial support for Views

justinvp opened the pull request:
This PR adds initial support for views, i.e. virtual child resources of
an owning resource. Note the `PULUMI_ENABLE_VIEWS_PREVIEW` envvar must
be set to `true` to turn on the feature (the feature flag will be
removed imminently).

Views are saved in the state with a `ViewOf` property, which is the URN
of the owning resource. The view itself is a child resource of the
owning resource.

Views are managed by a provider publishing view steps during CRUD
operations of the owning resource. `ResourceStepPre` events are emitted
for each view step published. The view steps are executed during
execution of owning resource step's `Apply` method.

Part of https://github.com/pulumi/pulumi/issues/19240

justinvp commented on pkg/resource/deploy/deployment.go:
```diff
@@ -314,6 +314,10 @@ type Deployment struct {
+	// a channel to post steps back to the step generator.
+	resourceStatusEvents <-chan SourceEvent
+	// the resource status server.
+	resourceStatus *resourceStatusServer
```
This is here out of convenience. It probably makes more sense on the deployment executor than the deployment, but all the existing steps already have a reference to the deployment, so it's convenient to access from here rather than updating all the steps to take in `resourceStatus`.

justinvp commented on pkg/resource/deploy/resource_status.go:
```diff
@@ -0,0 +1,291 @@
+
+type tokenInfo struct {
+	urn     resource.URN
+	mu      sync.Mutex
+	batches [][]Step
```
I did not originally intend to track things like this. Really, I just want to track `map[token]urn`. I added this extra information, more as a temporary workaround to a problem I hit.

View steps are published inside a provider operation, e.g. create. I was running into a situation where the view resources were being saved into the snapshot _before_ the owning resource's state, resulting in snapshot integrity error. As a temporary workaround to get things "working," I am saving the steps here. Before calling `prov.Create`, `ReserveToken` is called to get a token, any published steps to that token are then saved here (but not "published" yet), and then in the `Step.Apply`'s completion function (which is called _after_ the owning resource has finished), it calls `ReleaseToken` which actually publishes the steps for the views. The downside is the views show up all at once after the resource operation is finished.

I'm planning to change this. What I want to have happen is as soon as the view steps are published to the ResourceStatus service, the `resourcePreEvent`s are published. Then after the provider operation is finished, the views get their `resOutputsEvent`s published and written to state.

As of right now a view step represents an already completed operation (both pre & post events). In the future, we may want to allow sending "pre" steps first (which only results in a `resourcePreEvent`), followed by the finalized "pre & post" steps (which only results in a `resOutputsEvent`). We could do it by adding to the handshake and adding a field to the ViewStep to indicate it is a "pre" step.

It does beg the question of whether sending steps is the right protocol, or if it should be the events themselves, e.g. `resourcePreEvent` and `resOutputsEvent`, as we had talked about in the design discussion. It's possible this could simplify the implementation. Rather than dealing with view steps flowing through, we'd just have to flow the events.

Now that I'm thinking of it, even with the current step-based protocol, I could turn the view steps immediately into their pre/post events, and not have them even play a role as "virtual" steps in the engine. I think I'll try that out to see if it simplifies things at all.

lunaris replied:
This feels like the best thing to chew on and work out. Everything else here looks really sensible to me, and I think any of the approaches in your comment could work, but something more explicitly pre-/post- could be cool to get the updates flowing as soon as possible etc. Exciting!

lunaris replied:
Getting ahead of myself, but we might not even need to worry about handshake -- none of this is in use yet so we are free to change the protocol a bit to make it work I think (no released engine is sending `true` by default for `supportsViews`, right?)

justinvp replied:
From a cursory try of just using events in the implementation, it doesn't seem to make much of a difference implementation-wise.

Protocol-wise, I think what we have settled on around _steps_ rather than _events_ is still fine. If we switched the protocol to publishing events rather than steps, those events include the associated step metadata as part of their payload, so you're still passing the same information either way, but with events, in some sense it's more verbose.

I do think we should consider adding an `ExecutionPhase` enum (open to better naming) and field to `ViewStep`, something like:

```proto
enum ExecutionPhase {
    PRE_AND_POST = 0;
    PRE = 1;
    POST = 2;
}
```

- `PRE_AND_POST` is the current (default) behavior (which is all that's needed currently for TF modules). It means we'll raise both a `resourcePreEvent` and `resOutputsEvent` ("post") event for the step.
- `PRE` only results in a `resourcePreEvent` being raised.
- `POST` only results in a `resOutputsEvent` being raised.

If you had info about a view early in the process, you could send `PRE` steps first, and then finalize with `POST` steps as the end.

If you don't have any early info about a view, you'd just do the default `PRE_AND_POST` (or not even set it since 0 is the default).

It's easy to add support for this enum now, if we want. And, yeah, if we add it now, it wouldn't require any handshake changes. If we deferred it until sometime later, after TF modules, we'd probably need a handshake.

justinvp commented on pkg/resource/deploy/step.go:
```diff
@@ -1785,3 +1919,68 @@ func (s *DiffStep) Fail() {
+// ViewStep isn't really a step like a normal step. It's really a virtual step for a view resource. The step itself
+// doesn't perform any operations against a provider, it's used to communicate the steps that were taken for the
+// view resource for display purposes.
+type ViewStep struct {
```
An alternative approach would be folding this into the existing steps (e.g. `CreateStep`) and adding a few view-specific fields, including one that indicates it is a view step.

lunaris commented on pkg/resource/deploy/step_generator.go:
```diff
@@ -1921,6 +1957,46 @@ func (sg *stepGenerator) GenerateDeletes(targetsOpt UrnTargets, excludesOpt UrnT
+			views = append(views, view)
+		}
+	}
+	return views
```
Nit: Is `deployment.prev.Resources` mutated during a deployment? If not, maybe we can build a `map[URN][]plugin.View`s once and use that instead of this (apologies if this is already on the TODO list and just not done for expediency of sharing here).

lunaris commented on pkg/resource/deploy/step_generator.go:
```diff
@@ -1890,6 +1914,18 @@ func (sg *stepGenerator) GenerateDeletes(targetsOpt UrnTargets, excludesOpt UrnT
+	{
+		filtered := []Step{}
+		for _, step := range steps {
+			if step.Old() == nil || step.Old().ViewOf == "" {
```
This is filtering _out_ view steps, right? I wonder if inverting the condition might be clearer, or maybe it's fine with a comment.

lunaris commented on pkg/resource/deploy/step_generator.go:
```diff
@@ -1921,6 +1957,46 @@ func (sg *stepGenerator) GenerateDeletes(targetsOpt UrnTargets, excludesOpt UrnT
+}
+
+// getViews returns the set of views for a given URN.
+func (sg *stepGenerator) getViews(urn resource.URN) []plugin.View {
```
`getOldViews`, maybe?

lunaris commented on pkg/resource/deploy/resource_status.go:
```diff
@@ -0,0 +1,291 @@
+	stateType := tokens.Type(state.GetType())
+	stateURN := rs.deployment.generateURN(viewOf, stateType, state.GetName())
+
+	// TODO parenting
```
Underrated comment

justinvp commented on pkg/resource/deploy/deploytest/resourcestatus.go:
```diff
@@ -0,0 +1,233 @@
+// Copyright 2025, Pulumi Corporation.
```
I was just being consistent with the rest of the names in the directory

<img width="210" alt="Screenshot 2025-06-03 at 3 27 54 PM" src="https://github.com/user-attachments/assets/0f737a2a-456a-42cc-9ea1-968d43d87627" />

justinvp commented on pkg/backend/snapshot.go:
```diff
@@ -191,7 +191,7 @@ type sameSnapshotMutation struct {
 // mustWrite returns true if any semantically meaningful difference exists between the old and new states of a same
 // step that forces us to write the checkpoint. If no such difference exists, the checkpoint write that corresponds to
 // this step can be elided.
```
Eh, not sure what else to say so left this one as-is.

justinvp commented on pkg/resource/deploy/deployment.go:
```diff
@@ -445,12 +454,33 @@ func buildResourceMap(prev *Snapshot, preview bool) ([]*resource.State, map[reso
+	}
+	if res.Parent != "" && res.Parent != res.ViewOf {
+		view.ParentType = res.Parent.Type()
+		view.ParentName = res.Parent.Name()
```
All views are explicitly parented to the owning resource, currently. When parenting is implemented, you'll be able to create nested hierarchies under the owning resource, e.g. so a view can be a parent of another view, all under the same owning resource.

This `plugin.View` shape is what we send to the provider CRUD operations for OldViews, and it presents a simpler shape of the view. When ParentType/ParentName aren't set, the view is implicitly parented to the owning resource. If it is set, it'll be parented to some other view under the owning resource.

justinvp commented on pkg/resource/deploy/step.go:
```diff
@@ -37,7 +37,7 @@ import (
 // functions are to be called when the engine has fully retired a step. You
 // _should not_ modify the resource state in these functions -- doing so will
 // race with the snapshot writing code.
```
Reverted the signature, so it is back to not returning an error.

justinvp commented on pkg/resource/deploy/step.go:
```diff
@@ -349,7 +367,11 @@ func (s *CreateStep) Apply() (resource.Status, StepCompleteFunc, error) {
+	complete := func() error {
+		err := s.deployment.resourceStatus.ReleaseToken(resourceStatusToken)
+		s.reg.Done(&RegisterResult{State: s.new})
+		return err
```
No longer relevant. Reverted.

justinvp commented:
Thanks for the feedback! I've addressed all the comments, everything besides Fraser's comment about synchronization. Assuming CI is green and there isn't any other feedback, I think I'll go ahead and merge this in the morning, and address any needed synchronization as a follow-up fix.
